Add old files

s9latimm opened the pull request.

vulder committed:
Adds simple readme

vulder committed:
Adds basic gitignore

vulder committed:
Adds initil build with empty files

vulder committed:
Adds docs batch

vulder committed:
Test travis config

vulder committed:
Lower cmake version for travis

vulder committed:
Travis finds build error - Check

vulder committed:
Expand build matrix

vulder committed:
Allow llvm-9

vulder committed:
Add minimal llvm selector

vulder committed:
Better llvm version selection

vulder committed:
Other llvm version selector

vulder committed:
trigger travis

vulder committed:
Check if optional

vulder committed:
Test empty var

vulder committed:
Test unittest setup

vulder committed:
Fixes test bug

vulder committed:
Adds clang-tidy to travis config

vulder committed:
Add second clang tidy lookup

vulder committed:
Fixate clang tidy version to 10

vulder committed:
Travis update

vulder committed:
test travis

vulder committed:
Fixes tidy issues

vulder committed:
Adds Overloaded helper

vulder committed:
Adds travis badge

vulder committed:
Removes legacy travis option

s9latimm committed:
Add old files

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
-#ifndef VARA_FEATURE_FEATURE_H
-#define VARA_FEATURE_FEATURE_H
+#ifndef VARA_FEATUREMODEL_FMFEATURE_H
+#define VARA_FEATUREMODEL_FMFEATURE_H
```
defines do not match HEADER_GUARD layout

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
+#define VARA_FEATUREMODEL_FMFEATURE_H
 
 #include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
+#include <variant>
 
-namespace vara::feature {
+#include "vara/Feature/Relationship.h"
```
Fix include order:
1) own headers
\n
2) external headers
\n
3) system headers

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
 
-class Feature {
-  llvm::SmallVector<int, 8> Data;
+using std::string;
```
```suggestion
using std::string;

```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
-class Feature {
-  llvm::SmallVector<int, 8> Data;
+using std::string;
+namespace vara {
```
As in the example header, all feature lib stuff should life in it's own namespace.

`namespace vara:feature:`

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
+//===----------------------------------------------------------------------===//
+
+/// \brief Base class for components of \a FeatureModel.
+class FMFeature {
```
We still call this Feature, with the correct namespace there will not be a name clash

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
+  using feature_iterator = typename FeatureListType::iterator;
+
+private:
+  string Name;
+
+  llvm::SmallVector<FMFeature *, 1> Parents;
+
+  FeatureListType Children;
+
+  FeatureListType Excludes;
+
+  FeatureListType Implications;
+
+  FeatureListType Alternatives;
+
+  RelationshipListTy Relationships;
+
+  bool Opt;
```
```suggestion
  string Name;
  llvm::SmallVector<FMFeature *, 1> Parents;
  FeatureListType Children;
  FeatureListType Excludes;
  FeatureListType Implications;
  FeatureListType Alternatives;
  RelationshipListTy Relationships;
  bool Opt;
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
+/// \brief Base class for components of \a FeatureModel.
+class FMFeature {
+public:
+  using FeatureListType = typename llvm::SmallVector<FMFeature *, 3>;
+
+  using RelationshipListTy =
+      typename llvm::SmallVector<std::unique_ptr<Relationship<FMFeature>>, 3>;
+
+  using feature_iterator = typename FeatureListType::iterator;
```
```suggestion
  using FeatureListType = typename llvm::SmallVector<FMFeature *, 3>;
  using RelationshipListTy =
      typename llvm::SmallVector<std::unique_ptr<Relationship<FMFeature>>, 3>;
  using feature_iterator = typename FeatureListType::iterator;
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,26 +1,143 @@
+    Alternatives.push_back(Alternative);
+  }
+
+  using relationship_iterator = RelationshipListTy ::iterator;
+
```
move using up to the other using decls at the beginning

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+#ifndef VARA_FEATUREMODEL_FEATUREMODEL_H
+#define VARA_FEATUREMODEL_FEATUREMODEL_H
```
same

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+#ifndef VARA_FEATUREMODEL_FEATUREMODEL_H
+#define VARA_FEATUREMODEL_FEATUREMODEL_H
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/GraphWriter.h"
+
+#include "vara/Feature/Feature.h"
```
same

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+
+#include "vara/Feature/Feature.h"
+
+using std::string;
```
```suggestion
using std::string;

```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+#include "vara/Feature/Feature.h"
+
+using std::string;
+namespace vara {
```
same

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+
+class FeatureModel {
+public:
+  using FeatureMapTy = llvm::StringMap<std::unique_ptr<FMFeature>>;
+
+  using ConstraintsTy = std::vector<std::vector<std::pair<FMFeature *, bool>>>;
```
```suggestion
  using FeatureMapTy = llvm::StringMap<std::unique_ptr<FMFeature>>;
  using ConstraintsTy = std::vector<std::vector<std::pair<FMFeature *, bool>>>;
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+  using ConstraintsTy = std::vector<std::vector<std::pair<FMFeature *, bool>>>;
+
+private:
+  string Name;
+
+  FeatureMapTy Features;
+
+  ConstraintsTy Constraints;
+
+  FMFeature *Root;
```
```suggestion
  string Name;
  FeatureMapTy Features;
  ConstraintsTy Constraints;
  FMFeature *Root;
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+  FMFeature *Root;
+
+public:
+  FeatureModel(string VM, FeatureMapTy &Features, ConstraintsTy &Constraints)
+      : Name(std::move(VM)), Features(std::move(Features)),
+        Constraints(std::move(Constraints)), Root(Features["root"].get()) {}
```
```suggestion
  FeatureModel(string VM, FeatureMapTy Features, ConstraintsTy Constraints)
      : Name(std::move(VM)), Features(std::move(Features)),
        Constraints(std::move(Constraints)), Root(Features["root"].get()) {}
```

If we move the objects into our own class we need to take the parameters by value

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -0,0 +1,222 @@
+
+  template <typename GraphWriter>
+  static void addCustomGraphFeatures(vara::FeatureModel *FM, GraphWriter &W) {
+    llvm::SmallSet<std::pair<vara::FMFeature *, vara::FMFeature *>, 10> SkipE,
+        SkipI, SkipA;
```
Could you please spell out the names, E/I/A makes it quite hard to read for me.
`SkipExcludes` would make it easier for me :grin:

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,81 @@
+#ifndef VARA_TMODEL_RELATIONSHIP_H
+#define VARA_TMODEL_RELATIONSHIP_H
```
same

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,81 @@
+#ifndef VARA_TMODEL_RELATIONSHIP_H
+#define VARA_TMODEL_RELATIONSHIP_H
+
+#include <sstream>
```
```suggestion
#include <sstream>

```

vulder commented on include/vara/Feature/XmlParser.h:
```diff
@@ -0,0 +1,73 @@
+#include "vara/Feature/FeatureModel.h"
+
+#include "libxml/parser.h"
+#include "libxml/tree.h"
+#include <memory>
```
```suggestion
#include "libxml/tree.h"

#include <memory>
```

vulder commented on include/vara/Feature/XmlParser.h:
```diff
@@ -0,0 +1,73 @@
+  using constXmlCharPtr = const xmlChar *;
+
+private:
+  std::string DocPath;
+
+  std::string DtdPath;
+
+  std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)> Dtd;
+
+  std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> Doc;
+
+  std::string VM;
+
+  FeatureModel::FeatureMapTy Features;
+
+  FeatureModel::ConstraintsTy Constraints;
+
+  std::vector<std::pair<std::string, std::string>> RawEdges;
+
+  std::vector<std::pair<std::string, std::string>> RawExcludes;
+
+  std::vector<std::vector<std::pair<std::string, bool>>> RawConstraints;
```
```suggestion
  std::string DocPath;
  std::string DtdPath;
  std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)> Dtd;
  std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> Doc;
  std::string VM;
  FeatureModel::FeatureMapTy Features;
  FeatureModel::ConstraintsTy Constraints;
  std::vector<std::pair<std::string, std::string>> RawEdges;
  std::vector<std::pair<std::string, std::string>> RawExcludes;
  std::vector<std::vector<std::pair<std::string, bool>>> RawConstraints;
```

vulder commented on lib/Feature/CMakeLists.txt:
```diff
@@ -11,3 +12,8 @@ set(LLVM_LINK_COMPONENTS
 )
+
+target_link_libraries(VaRAFeature
+  PUBLIC
```
```suggestion
  LINK_PUBLIC
```

vulder commented on lib/Feature/Feature.cpp:
```diff
@@ -1,14 +1,77 @@
 #include "vara/Feature/Feature.h"
 
-#include <iostream>
+namespace vara {
```
same

vulder commented on lib/Feature/XmlParser.cpp:
```diff
@@ -0,0 +1,267 @@
+#include <iostream>
+#include <regex>
+
+namespace vara {
```
same

vulder commented on lib/Feature/XmlParser.cpp:
```diff
@@ -0,0 +1,267 @@
+        P.second->addRelationship(std::make_unique<Mandatory<FMFeature>>(C));
+      }
+    }
+    // TODO relationships: or and xor
```
Should this be fixed? If you plan to fix it later please create an issue and write the TODO like this:
`TODO (se-passau/VaRA#42): do stuff to bar`

Otherwise, I'll always lose track of the TODOs :grinning:

vulder reviewed (changes requested):
Mostly found small syntactic nits, otherwise this looks good, as before :).

One thing that I found, if we use llvm::errs()  instread of std::err we could use llvm::formatv to make formatting ouf output easier.

s9latimm committed:
Merge branch 'vara-dev' into f-AddFiles

s9latimm committed:
Resolve

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/Relationship.h
M	include/vara/Feature/XmlParser.h
M	lib/Feature/CMakeLists.txt
M	lib/Feature/XmlParser.cpp

s9latimm committed:
Update namespace

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/Relationship.h
M	include/vara/Feature/XmlParser.h
M	lib/Feature/Feature.cpp
M	lib/Feature/XmlParser.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Reformat

M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Add travis libxml2 install

M	.travis.yml

s9latimm committed:
Fix cmake

M	CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp

vulder commented on CMakeLists.txt:
```diff
@@ -73,7 +73,7 @@ if (NOT VARA_FEATURE_IN_TREE)
   else()
     message(STATUS "Not set")
-    find_package(LLVM 10 REQUIRED CONFIG)
+    find_package(LLVM 9 REQUIRED CONFIG)
```
We require and support by default LLVM 10 if someone wants to build an older version, cmake offers a extra flag to allow older LLVM version. The readme explains how to build with older versions, but we do not guarantee that everything will work.

s9latimm committed:
Fix test

M	CMakeLists.txt
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/Feature.cpp

s9latimm committed:
Fix tidy

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Fix move

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/XmlParser.cpp
